File: examples/cpp/quick_online.cpp
#include "rclib/Model.h"
#include "rclib/readouts/RlsReadout.h"
#include "rclib/reservoirs/RandomSparseReservoir.h"

#include <Eigen/Dense>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <memory>

int main() {
  // 1. Data preparation (sine wave prediction)
  int n_samples = 1000;
  Eigen::VectorXd t = Eigen::VectorXd::LinSpaced(n_samples, 0, 100);
  Eigen::MatrixXd data = t.unaryExpr([](double v) { return std::sin(v); });

  Eigen::MatrixXd X = data.topRows(n_samples - 1);
  Eigen::MatrixXd Y = data.bottomRows(n_samples - 1);

  // 2. Initialization of ESN and RLS readout
  Model model;
  model.addReservoir(std::make_shared<RandomSparseReservoir>(500, 0.9));
  model.setReadout(std::make_shared<RlsReadout>(0.99, 1.0, true));

  // 3. Online learning (update weights sample by sample)
  for (int i = 0; i < X.rows(); ++i) {
    model.partialFit(X.row(i), Y.row(i));
  }

  // 4. Prediction and evaluation
  Eigen::MatrixXd Y_pred = model.predict(X);
  double mse = (Y - Y_pred).squaredNorm() / Y.rows();
  std::cout << "MSE: " << std::scientific << std::setprecision(2) << mse << std::endl;

  return 0;
}

File: examples/python/quick_online.py
"""Quick online learning example."""

from __future__ import annotations

import numpy as np
from rclib import ESN, readouts, reservoirs


def main() -> None:
    """Run the quick online learning example."""
    # 1. Data preparation (sine wave prediction)
    t = np.linspace(0, 100, 1000)
    data = np.sin(t).reshape(-1, 1)
    x, y = data[:-1], data[1:]

    # 2. Initialization of ESN and RLS readout
    model = ESN()
    model.add_reservoir(reservoirs.RandomSparse(n_neurons=500, spectral_radius=0.9))
    model.set_readout(readouts.Rls(lambda_=0.99, delta=1.0, include_bias=True))

    # 3. Online learning (update weights sample by sample)
    for i in range(len(x)):
        model.partial_fit(x[i : i + 1], y[i : i + 1])

    # 4. Prediction and evaluation
    y_pred = model.predict(x)
    mse = np.mean((y - y_pred) ** 2)
    print(f"MSE: {mse:.2e}")


if __name__ == "__main__":
    main()
